### PR TITLE
UI Overhaul: Improve pipeline editor performance

### PR DESCRIPTION
**Pull request by mausworks**
## Description

This PR overhauls a few performance critical components in the pipeline editor.
It does so by a few large-sweeping changes in the codebase.

TL;DR: The performance improvements are largely about avoiding object/array allocations in performance critical sections of the code, and trying to pass positions/points by reference when possible.

### Position's & Offset's are out: Point2D is in :fire: 

We used `Position` (`{ x: number; y: number }`), `Offset` (`{top: number, left: number}`) and `Point2D` (`[number, number]`) inconsistently throughout the code base. This meant we had to do translations between them, which (in performance critical scenarios) can cause bottlenecking.

By consistently using `Point2D` throughout the codebase, and by converting to them as early as possible, we can simply pass arrays as references between functions without having to do any conversion.

### Improved performance in the mouse tracker :computer_mouse: 

Previously our mouse tracker tracked these properties:
- client position
- client delta
- previous client position
- scaled position (relative to zoom)
- scaled delta (relative to zoom)

It also did so by creating a new `Position` object on each update for each property, instead of mutating the X/Y positions.
This caused the GC to go wild after a while, causing stuttering and lag.

The new mouse tracker only keeps track of the client position and delta from the previous position. It also mutates these properties for minimal performance impact. The "previous" position property is not used or needed, and the scaled positions can now be retrieved by using `dividePoint(getMousePoint(), scaleFactor)`. 

The `getMousePoint` function is available ~~from the `useTrackMouse` hook~~ in `@/utils/mouse` that can be used "cost-free" in any component.

#### Why an array instead of an object?

I had to pick one. I naturally wanted to pick the one with the best performance, but like with most things in JavaScript, the answer to which one is more performant "depends".

Objects _can_ have better performance in hot code paths. V8 may choose to cache `{ x: number, y: number }`  as a class, since it realizes it's a common object, but it's not always consistent. You can compare the performance of allocating arrays vs [objects in this codepen](http://jsfiddle.net/ryanwheale/HbHxv/4/). Clicking either of the buttons generates 1 million objects or arrays.
If you click the "Test Objects" once or twice, you will see a much longer duration compared to "Test Arrays" which is always consistent (for me around ~39-45ms). "Test Objects" sometimes outperforms "Test Arrays" (for me around ~35-55ms), but only if I spam the "Test Objects" button for some duration. I'd much rather have consistent performance than hope V8 optimizes my code oppropriately—and for 2D points, I think tuples are an excelent choice anyways.

### Other optimizations

I have tried (wherever possible) to re-use 2D-point references, instead of destructuring them and reconstructing them. Things that previously had `x` and `y` properties now generally have an `origin`-property instead: a `Point2D`. This also makes it easier to use the new geometry abstractions with them: `subtractPoints`, `addPoints`, `dividePoint` and `multiplyPoint`. A some of the and-rolled functions have been replaced with these generic ones.
 
## Checklist

- [x] I have manually tested my changes and I am happy with the result.

### CODE REVIEW

**Comment by yannickperrenet** [comment]
I love it! (Based on what is described in the PR description.) :heart:

**Review by iannbing — APPROVED** [review]
The changes look good to me 👍, and I think in theory it improves performance (I didn't do deep profiling, hard to tell the difference for me 🙂, comparing to `dev`). Thanks for explaining how you approach the issue you found. A great lesson for me  😄.

Happy to approve this PR as it's a nice improvement on the javascript level (only left some minor syntax improvement suggestions.). Yet I think the main bottleneck is at the React level. For example, when creating/resizing the selection rectangle, File Manager is also re-rendered. In my experience, solving the React issue usually brings tangible performance boosts. Eager to see the improvements on that side 😄.

**Comment by ncspost** [comment]
Great work @mausworks 

Panning is 🔥 

Zooming feels buttery. I still think the acceleration on it could be slightly increased tho.

File: services/orchest-webserver/client/src/hooks/useTrackMouse.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import { Point2D } from "@/utils/geometry";
+import { useCallback, useEffect, useRef } from "react";
```
[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/hooks/useTrackMouse.ts:2]: Sure, this is just habit! I'll fix. Any particular reason we like to use `React.useHook`?

[the remark this answers is not in the document]
Comment (mausworks) [reply services/orchest-webserver/client/src/hooks/useTrackMouse.ts]: I actually went with a different solution in the end. We don't really need a hook for the mouse tracker at all. We just need one global `mousemove` handler, and that's it. So I made it a singleton. See the commit message for [791f9e2](https://github.com/orchest/orchest/pull/1204/commits/791f9e24fe99bfa3d46efacc4dac922e2897b664).

**Pull request merged**